Sources/Regex/Atom.swift

```diff
@@ -305,20 +308,51 @@ extension Atom.EscapedBuiltin {
 }
 
 extension Atom {
-  // TODO: Hamish, I believe you have a formulation of this and have
-  // thought through the parsing a whole lot more. This is just what
-  // I have at the time, but let's get something better for the AST
-  // and parser support.
-  public enum Prop: Hashable {
-    case gc(Unicode.GeneralCategory)
-    case pcreSpecial(PCRESpecialCategory)
+  public struct CharacterProperty: Hashable {
+    public var kind: Kind
+    public var isInverted: Bool
+
+    public init(_ kind: Kind, isInverted: Bool) {
+      self.kind = kind
+      self.isInverted = isInverted
+    }
+  }
+}
+
+extension Atom.CharacterProperty {
+  public enum Kind: Hashable {
+    /// Matches any character, equivalent to Oniguruma's '\O'.
+    case any
+
+    // The inverse of 'Unicode.ExtendedGeneralCategory.unassigned'.
+    case assigned
+
+    /// All ascii characters U+00...U+7F
+    case ascii
+
+    /// A general category property.
+    case gc(Unicode.ExtendedGeneralCategory)
+
+    /// Binary character properties. Note that only the following are required
+    /// by UTS#18 Level 1:
+    /// - Alphabetic
+    /// - Uppercase
+    /// - Lowercase
+    /// - White_Space
+    /// - Noncharacter_Code_Point
+    /// - Default_Ignorable_Code_Point
+    case binary(Unicode.BinaryProperty, value: Bool)
+
+    /// Character script and script extensions.
     case script(Unicode.Script)
+    case scriptExtension(Unicode.Script)
 
-    // TODO: replace me
-    case propCheck(PropName, PropValue)
+    /// Some special properties implemented by PCRE and Oniguruma.
+    case pcreSpecial(PCRESpecialCategory)
+    case onigurumaSpecial(OnigurumaSpecialProperty)
```

Comment:
Does Oniguruma have anything special beyond Unicode?

Reply:
I don't think so, I'm pretty sure the only properties remaining are those for the `Block` Unicode property (though Oniguruma prefixes them with `In_`, presumably as it would otherwise cause ambiguities with the script properties), but I need to go through and double check that. If that is the case, I think we should probably parse out the block properties along with their Unicode aliases and Oniguruma `In_` prefixes (and decide when you need to disambiguate with `Block=`), but for now I figured we should at least parse out the bits that Oniguruma accepts.